src/shims/mod.rs

```diff
@@ -44,11 +40,13 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
         // Try to see if we can do something about foreign items.
         if this.tcx.is_foreign_item(instance.def_id()) {
-            // An external function that we cannot find MIR for, but we can still run enough
+            // An external function that we (possibly) cannot find MIR for, but we can still run enough
             // of them to make miri viable.
-            this.emulate_foreign_item(instance.def_id(), args, dest, ret)?;
-            // `goto_block` already handled.
-            return Ok(None);
+            // This will be Ok(None) if we're emulating the intrinsic entirely within Miri (no need
+            // to run extra MIR), and Ok(Some(body)) if we found an implementation for the
+            // foreign function
+            // Any needed call to `goto_block` will be performed by `emulate_foreign_item`
```

Comment:
```suggestion
            // Any needed call to `goto_block` will be performed by `emulate_foreign_item`.
```

Reply:
(Most of your comments lack trailing full stops, I am not pointing out all of them but it would be good to be consistent about having them.)